### PR TITLE
fix: hardhat_setCode error handling

### PR DESCRIPTION
**Pull request by vbar**
# What :computer: 
Replaced `hardhat_setCode` implementation to not panic on code parsing.

# Why :hand:
The code argument is user-controlled, and users shouldn't be able to crash the node
just by calling its API. At #252, they're complaining it can happen even accidentally.

# Evidence :camera:
Added test:
```
  hardhat_setCode
    ✔ Should set code at an address (279ms)
    ✔ Should reject invalid code
    ✔ Should update code with a different smart contract (291ms)

```

### CODE REVIEW

File: src/node/in_memory.rs [hunk 1]
```diff
@@ -1518,11 +1518,14 @@ impl<S: ForkSource + std::fmt::Debug + Clone> InMemoryNode<S> {
...
-        let bytecodes: HashMap<U256, Vec<U256>> = vm
-            .get_last_tx_compressed_bytecodes()
-            .iter()
-            .map(|b| bytecode_to_factory_dep(b.original.clone()))
-            .collect();
+        let mut bytecodes = HashMap::new();
+        for b in vm.get_last_tx_compressed_bytecodes().iter() {
+            let hashcode = match bytecode_to_factory_dep(b.original.clone()) {
+                Ok(hc) => hc,
+                Err(error) => { return Err(error.to_string()); }
```
[the remark this answers is not in the document]
Comment (vbar) [reply src/node/in_memory.rs]: Well, that can be done (in https://github.com/matter-labs/era-test-node/commit/fdb17ddc8ac70f4d9a35cc01f1c181b14bf3a2d2 ), but is it worth using custom colors here? The other call to `tracing::error!` in this file uses default colors...

File: e2e-tests/test/hardhat-apis.test.ts [hunk 2]
```diff
@@ -118,6 +118,22 @@ describe("hardhat_setCode", function () {
...
+
+      const address = "0x1000000000000000000000000000000000001111";
+      const artifact = await deployer.loadArtifact("Return5");
+      const contractCode = [...ethers.utils.arrayify(artifact.deployedBytecode)];
+      const shortCode = contractCode.slice(0, contractCode.length - 1);
+
+      await provider.send("hardhat_setCode", [address, shortCode]);
+    };
+
+    await expectThrowsAsync(action);
```
Comment (MexicanAce) [inline e2e-tests/test/hardhat-apis.test.ts:134]: This requires a second parameter, the `errorMessage` string to be a substring in the expected error to be thrown

Comment (MexicanAce) [reply e2e-tests/test/hardhat-apis.test.ts:134]:
```suggestion
    await expectThrowsAsync(action, "bytes must be divisible by 32");
```

Comment (MexicanAce) [reply e2e-tests/test/hardhat-apis.test.ts]: Doesn't look like I chose the correct error message 😅

Comment (vbar) [reply e2e-tests/test/hardhat-apis.test.ts]: The error message is correct - it's just that the server looses it when converting the error through `Web3Error::InternalError`... On the other hand, there is no reason it should convert errors through `Web3Error::InternalError` (except it had been implemented that way) - with https://github.com/matter-labs/era-test-node/pull/293/commits/8b7b5e1639a0c0b4bdc1eb7db833e011302cffff , the test passes unmodified. This change should probably be applied more widely... :-)

**Review by MexicanAce — APPROVED** [review]
LGTM

**Pull request merged**